### PR TITLE
[#2843] Mark initially assigned partitions for offset commit

### PR DESCRIPTION
**Pull request by calohmn**
This is for #2843.
When a partition gets assigned for the first time to the `AsyncHandlingAutoCommitKafkaConsumer`, the initial partition
offset gets committed on the first offset commit.
This makes sure that a record published during a following rebalance doesn't get skipped.

### CODE REVIEW

**Comment by calohmn** [comment]
@b-abel Can you have a look?

**Comment by b-abel** [comment]
> @b-abel Can you have a look?

Sure.

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java [hunk 1]
```diff
@@ -464,11 +514,12 @@ private synchronized void setCommittedOffsets(final Map<TopicPartition, OffsetAn
      */
     class TopicPartitionOffsets {
 
+        private static final long UNDEFINED_OFFSET = -2;
         private final TopicPartition topicPartition;
         private final Deque<OffsetsQueueEntry> queue = new LinkedList<>();
 
-        private long lastSequentiallyCompletedOffset = -1;
-        private long lastCommittedOffset = -1;
```
Comment (b-abel) [inline clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java:471]: Why are the defaults changed from `-1` to `-2`?

Comment (calohmn) [reply clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java:471]: The `TopicPartitionOffsets` inner class works with offsets that refer to the position of a received/handled record.
The offset for the actual offset commit is `+1` of that value.
With this PR, we now have the situation that we explicitly want to do an offset commit with position 0. That means that `lastSequentiallyCompletedOffset` has to be set to `-1` in that case, meaning `-1` can't be used as the number referring to an unknown offset anymore.

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java [hunk 2]
```diff
@@ -312,10 +324,46 @@ protected void onRecordHandlerSkippedForExpiredRecord(final KafkaConsumerRecord<
     protected void onPartitionsAssignedBlocking(final Set<io.vertx.kafka.client.common.TopicPartition> partitionsSet) {
         final Consumer<String, Buffer> wrappedConsumer = getKafkaConsumer().asStream().unwrap();
```
[the remark this answers is not in the document]
Comment (calohmn) [reply clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java]: I've created a new `getUnderlyingConsumer()` method and used it here.

File: clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java [hunk 3]
```diff
@@ -312,10 +324,46 @@ protected void onRecordHandlerSkippedForExpiredRecord(final KafkaConsumerRecord<
...
         clearObsoleteTopicPartitionOffsets(wrappedConsumer.assignment());
+        // remove lastKnownCommittedOffsets entries belonging to deleted topics
+        if (topicPattern != null) {
+            final Set<String> subscribedTopicPatternTopics = getSubscribedTopicPatternTopics();
+            lastKnownCommittedOffsets.entrySet()
+                    .removeIf(entry -> !subscribedTopicPatternTopics.contains(entry.getKey().topic()));
+        }
+        if (!"earliest".equals(consumerConfig.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG)) && !partitionsSet.isEmpty()) {
+            // for each partition ensure an offset gets committed on the next commit if there possibly has never been a commit before;
```
[the remark this answers is not in the document]
Comment (calohmn) [reply clients/kafka-common/src/main/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumer.java:332]: Yes, I've extracted a new method here.

**Comment by b-abel** [comment]
LGTM

**Pull request merged**